core/dataset_manager.py

```diff
@@ -107,24 +108,51 @@ def _validate_data(self):
                         folder=folder
                     )
 
-    def bootstrap(self, mapping_filepath="core/datasets.yaml"):
-        if self.mappings:
-            return False
+    def get_mappings(self):
+        """
+        Returns the dataset mappings.
+        """
+        if not self._mappings:
+            raise Exception("Dataset Manager needs to be bootstrapped first.")
+        return self._mappings
+
+    def bootstrap(self):
+        """
+        Check if datasets.yml exists. If so, load it into the class (if it
+        hasn't been already). Otherwise, call _create_dataset_mappings
+
+        Returns True if mappings had to be created, and False otherwise 
+        (used for testing).
+        """
+        mapping_filepath = os.path.join(self.raw_filepath, "datasets.yml")
         if os.path.isfile(mapping_filepath):
-            with open(mapping_filepath, "r") as f:
-                self.mappings = yaml.load(f.readlines())
+            if not self._mappings:
+                with open(mapping_filepath, "r") as f:
+                    self._mappings = yaml.load(f.readlines())
```

Comment:
Since this will always call _create_dataset_mappings() which will always set the mappings, what's the point of doing this here? Is it the same or different? They should probably both be in one place. If it's duplicated in both places, I get that it's "defensive programming" but I also think this would make it harder to spot any errors. This method should assume that _create_dataset_mappings does exactly what it's supposed to do, and either call it or not call it but not try to do part of its job.

Reply:
Can you clarify what you mean? _create_dataset_mappings is only called when the `datasets.yaml` doesn't exist.

Reply:
On startup, datasets.yaml doesn't exist, right.
Then _create_dataset_mappings is called which sets self._mappings
So why is this again setting self._mappings?

Reply:
Oh, the mappings have to stay consistent. It'll make more sense if I explain with an example.

When data provider's ED Directories are pushed, identifiers are used to represent the datasets. Then the Notebook will use these identifiers to tell the service what datasets the model developer wants to train on. So let's say the service stopped running on the data provider's machine for whatever reason so that the mappings aren't loaded when it starts again. The mappings can't be regenerated (random strings), so we need to use the same mappings.

Tl;dr identifiers are non-deterministic so they need to be saved so the service knows what datasets to train on at a later point.

Reply:
Ok makes sense thanks for example.